content/posts/unpublished/ways-to-improve-be-api-performance/post.md

```diff
@@ -0,0 +1,304 @@
+  * Debugging and testing may require specialized frameworks and techniques.
+
+# Parallel processing
+
+## Use cases and advantages
+
+  During development of a service in microservice architecture the following situation may arise:
+  the task of the service is to query several different services and aggregate data fetched from them. The straightforward approach is to fetch data by querying those services one by one,
+  but if those calls are independent, we might speed up the process by executing the queries in parallel.
+  Let's consider an example where a service fetches data from Product, Payment, and Order services for further processing. Assume that it takes 200 ms, 100 ms and 150 ms accordingly. 
+  The comparison of these two approaches is shown in the diagram.
+
+![](assets/parallel.png)
+  
+  To implement the parallel approach, `CompletableFuture` may be used, CompletableFuture is a class introduced in Java 8 that allows us to write asynchronous, non-blocking code. Let's look at the example:
+  ```Java
+{
+    var productFuture=CompletableFuture.supplyAsync(()->ProductService.fetch(request),exec);
+    var paymentFuture=CompletableFuture.supplyAsync(()->PaymentService.fetch(request),exec);
+    var orderFuture=CompletableFuture.supplyAsync(()->OrderService.fetch(request),exec);
+
+    CompletableFuture<Void> allFutures=CompletableFuture.allOf(productInfoFuture,paymentFuture,orderFuture);
+
+    try {
+        allFutures.get(); // Waits until all tasks are completed
+    } catch(InterruptedException|ExecutionException e){
+        e.printStackTrace();
+    }
+
+    try {
+        requiredData.setProductInfo(productFuture.get());
+        requiredData.setPaymentInfo(paymentFuture.get());
+        requiredData.setOrderInfo(orderFuture.get());
+    } catch(InterruptedException|ExecutionException e){
+        e.printStackTrace();
+    }
+    return requiredData;
+}
+  ```
+  
+  Virtual threads are a new feature introduced in Java 21. Virtual threads are lightweight threads that reduce the effort of writing, maintaining, and debugging high-throughput concurrent applications.
+  Unlike regular thread, virtual threads does not entail any OS-level blocking. Let's try to implement the previous example using Virtual Threads.
+
+  ```Java
+{
+    Thread t1=startVirtualThread(()->requiredData.setProductInfo(ProductService.fetch(request)));
+    Thread t2=startVirtualThread(()->requiredData.setPaymentInfo(PaymentService.fetch(request)));
+    Thread t3=startVirtualThread(()->requiredData.setOrderInfo(OrderService.fetch(request)));
+
+    t1.start();t2.start();t3.start();
+    t1.join();t2.join();t3.join();
+    return requiredData;
+}
+  ```  
+
+  Spring Webclient is another tool we can use to make parallel service calls. WebClient uses an asynchronous, non-blocking solution provided by the Spring Reactive framework:
```

Comment:
good to mention that since java 21 using virtual threads is probably going to be the prefered approach given the complexity of implementing reactive components

Reply:
updated